Minio/DataModel/Select/SelectResponseStream.cs

```diff
@@ -0,0 +1,239 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+using Minio.Exceptions;
+
+using System;
+using System.Collections.Generic;
+
+using System.IO;
+using System.Runtime.Serialization.Formatters.Binary;
+using System.Xml.Serialization;
+using RestSharp;
+using System;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Force.Crc32;
+using System.Linq;
+using System.Collections.Generic;
+
+namespace Minio.DataModel
+{
+    [Serializable]
+    public class SelectResponseStream
+    {
+        // SelectResponseStream is a struct for selectobjectcontent response.
+        
+        public Stream Payload  { get ; set; }  
+
+        [XmlElement("Stats", IsNullable = false)]
+        public StatsMessage Stats { get; set; }
+
+        [XmlElement("Progress", IsNullable = false)]
+        public ProgressMessage Progress { get; set; }
+
+        private byte[] prelude = new byte[8];
+        private byte[] preludeCRC = new byte[4];
+        private byte[] messageCRC = new byte[4];
+        private byte[] headerValueLen = new byte[2];
+
+        private MemoryStream payloadStream = null;
+        private bool _isProcessing;
+        
+        public SelectResponseStream()
+        {
+        }
+
+        protected int ReadFromStream(byte[] buffer)
+        {
+            int read = -1;
+            if  (!this._isProcessing)
+            {
+               return read;
+            }
+            read = this.payloadStream.Read(buffer, 0, buffer.Length);
+            if (!this.payloadStream.CanRead)
+            {
+                this._isProcessing = false;
+            }
+            return read;
+        }
+        public SelectResponseStream(Stream s)
+        {
+            if (s != null) {
+                var _ms = new MemoryStream();
+                s.CopyTo(_ms);
+                this.payloadStream = _ms;
+                this.Payload = new MemoryStream();
+            }
+            this._isProcessing = true;
+            this.payloadStream.Seek(0, SeekOrigin.Begin);
+            this.start();
+        }
+        
+        private void start()
+        {
+            int numBytesRead = 0;
+            int msgOffset = 0;
+            while (_isProcessing)
+            {   
+                var n =  ReadFromStream(prelude);
+                numBytesRead += n;
+                n =  ReadFromStream(preludeCRC);
+                numBytesRead += n;
+                var inputArray = new byte[prelude.Length + preludeCRC.Length + 4];
+                System.Buffer.BlockCopy(prelude,0,inputArray,0, prelude.Length);
+                System.Buffer.BlockCopy(preludeCRC,0,inputArray,prelude.Length, preludeCRC.Length);
+
+                // write real data to inputArray
+                Crc32Algorithm.ComputeAndWriteToEnd(inputArray); // last 4 bytes contains CRC
```

Comment:
Here the `computeAndWriteToEnd` calculates the CRC for the first 8(prelude) + 4(preludeCRC) bytes and writes the CRC in the last 4 byte block. We should not included the preludeCRC in the CRC calculation.

let the inputArray be `var inputArray = new byte[prelude.Length + 4];` and let us copy the prelude alone to the inputArray and let `ComputeAndWriteToEnd` calculate the CRC for the first 8bytes (prelude) and write that in the last 4 byte block. We could then verify this calculated CRC in the last 4bytes with the preludeCRC that we got from the stream - (`n =  ReadFromStream(preludeCRC);`)

Reply:
fixed.